src/xen/xennet.c

```diff
@@ -99,60 +99,74 @@ struct xennet_dev {
 
     u32 tx_id;
-    boolean again = false;
     char *node;
-    do {
-        s = xenstore_transaction_start(&tx_id);
-        if (!is_ok(s))
-            return s;
-
-        node = "vifname";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "vif%d", xd->if_id);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "rx-ring-ref";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->rx_ring_gntref);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "tx-ring-ref";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->tx_ring_gntref);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "request-rx-copy";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "feature-rx-notify";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
-        if (!is_ok(s))
-            goto abort;
-
-        node = "event-channel";
-        s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->evtchn);
-        if (!is_ok(s))
-            goto abort;
-
-        s = xenstore_transaction_end(tx_id, false);
-        if (!is_ok(s)) {
-            /* XXX check for EAGAIN */
-            goto abort;
+    int retries = XENNET_INFORM_BACKEND_RETRIES;
+
+  again:
+    node = "transaction start";
+    s = xenstore_transaction_start(&tx_id);
+    if (!is_ok(s))
+        return s;
+
+    node = "vifname";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "vif%d", xd->if_id);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "rx-ring-ref";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->rx_ring_gntref);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "tx-ring-ref";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->tx_ring_gntref);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "request-rx-copy";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "feature-rx-notify";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", 1);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "event-channel";
+    s = xenstore_sync_printf(tx_id, xd->frontend, node, "%d", xd->evtchn);
+    if (!is_ok(s))
+        goto abort;
+
+    node = "transaction end";
+    s = xenstore_transaction_end(tx_id, false);
+    if (!is_ok(s)) {
+        value v = table_find(s, sym(errno));
+        if (v) {
+            if (!runtime_strcmp("EAGAIN", buffer_ref((buffer)v, 0))) {
+                deallocate_tuple(s);
+                if (retries-- == 0) {
+                    s = timm("result", "%s failed: transaction end returned EAGAIN after %d tries",
+                             __func__, XENNET_INFORM_BACKEND_RETRIES);
+                    goto abort;
+                }
+                goto again;
+            }
         }
-    } while (again);
+        goto abort;
```

Comment:
I think you're leaking the tuple here

Reply:
Thanks, that error handling was messed up. I actually took the opportunity to sneak in "timm_up," which will allow us to trivially build up chains of statuses after a failure (that roughly follow the call stack). More on that later - we can do a lot to take advantage of these status tuples.

Deallocating tuples in general requires a bit more work to make it so that any buffer (wrapped / not wrapped, shared data, etc.) can be safely deallocated. Maybe also loop detection would be in order, as there's nothing to prevent cycles from being formed.